lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -401,7 +403,95 @@ static void endLexicalLifetimeBeforeInst(AllocStackInst *asi,
   SILLocation loc =
       RegularLocation::getAutoGeneratedLocation(beforeInstruction->getLoc());
   builder.createEndBorrow(loc, values.borrow);
-  builder.emitDestroyValueOperation(loc, values.stored);
+  builder.emitDestroyValueOperation(loc, values.stored->getOwnershipKind() ==
+                                                 OwnershipKind::Guaranteed
+                                             ? values.copy
+                                             : values.stored);
+}
+
+/// Returns true if \p src is a function parameter or a projection of it.
+static bool shouldExtendLifetimeForStoreBorrowSource(SILValue src) {
+  if (isa<SILFunctionArgument>(src))
+    return false;
+  if (Projection::isObjectProjection(src)) {
+    return shouldExtendLifetimeForStoreBorrowSource(
+        cast<SingleValueInstruction>(src)->getOperand(0));
+  }
+  return true;
+}
+
+/// Extends the lifetime of \p sbi source by creating a copy and a borrow. i.e:
+/// Transform store_borrow %src to %asi:
+/// %copy = copy_value %src
+/// %borrow = begin_borrow %copy
+static StorageStateTracking<LiveValues>
+extendLifetimeForStoreBorrowSource(AllocStackInst *asi, StoreBorrowInst *sbi) {
+  assert(shouldExtendLifetimeForStoreBorrowSource(sbi->getSrc()));
+
+  BeginBorrowInst *bbi = nullptr;
+  CopyValueInst *cvi = nullptr;
+  SILBuilderWithScope::insertAfter(sbi, [&](SILBuilder &builder) {
+    SILLocation loc = RegularLocation::getAutoGeneratedLocation(sbi->getLoc());
+    cvi = builder.createCopyValue(loc, sbi->getSrc());
```

Comment:
The intention behind `store_borrow` is that there is no semantic copy of the stored value. Can we avoid generating the `copy_value`?

Reply:
Hmm, I was thinking other lifetime optimizations can easily eliminate this copy. 
If the `store_borrow` source is a `begin_borrow`, its scope maybe ending before other `load_borrow` from the source. Should I avoid optimizing for such store_borrow sources ?